### PR TITLE
Feature/configure from sql

### PR DESCRIPTION
**Pull request by mrmasterplan**
### Configuration from sql

When using sql statements to create and manage tables, it can be good to keep all
information referring to a table within one file. Therefore, the configurator employs 
a sql parsing library to extract table details directly from the CREATE statements. 
It can be used like this:

```python
from . import my_sql_folder
Configurator().add_sql_resource_path(my_sql_folder)
```

A sql CREATE statement already contains all information that the configurator needs 
except the key, aka. table_id, that shall be used to refer to the table.
This key needs to be added to the sql in the form of a comment with a magic prefix
`"-- atc.Configurator "` (note the spaces. not case-sensitive.) See the following 
example:
```sparksql
-- atc.Configurator key: MySparkDb
CREATE DATABASE IF NOT EXISTS `my_db1{ID}`
COMMENT "Dummy Database 1"
LOCATION "/tmp/foo/bar/my_db1/";

-- atc.Configurator key: MyDetailsTable
CREATE TABLE IF NOT EXISTS `my_db1{ID}.details`
(
  {MyAlias_schema},
  another int
  -- comment with ;
)
USING DELTA
COMMENT "Dummy Database 1 details"
LOCATION "/{MNT}/foo/bar/my_db1/details/";

-- pure configurator magic in this statement
-- atc.Configurator key: MyAlias
-- atc.Configurator alias: MySqlTable
;


-- ATC.CONFIGURATOR key: MySqlTable
-- atc.Configurator delete_on_delta_schema_mismatch: true
CREATE TABLE IF NOT EXISTS `{MySparkDb}.tbl1`
(
  a int,
  b int,
  c string,
  d timestamp
)
USING DELTA
COMMENT "Dummy Database 1 table 1"
LOCATION "/{MNT}/foo/bar/my_db1/tbl1/";
```

The example is quite complex and demonstrates a number of features:
- the magic tag "-- ATC.CONFIGURATOR " is case-insensitive
- the key has to be specified as `-- ATC.CONFIGURATOR key: MyKey`.
- any other attribute can be specified after the same magic tag and will be 
  available through the configurator. In fact, all comments with the magic tag will 
  be collected together and will be interpreted as a `yaml` document, so more 
  complicated structures are also possible.
- all python string substitutions that the configurator offers are also available 
  here in the sql configuration.
- string substitution is available inside the schema, this allows schemas to be 
  constructed from other schemas.

### CODE REVIEW

File: src/atc/configurator/_parse_sql.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+    yaml_document = "\n".join(collected_lines)
+    return yaml.load(yaml_document, Loader=yaml.FullLoader) or {}
+
+
+@dataclass
+class StatementBlocks:
+    schema: str = None
+    using: str = None
+    options: str = None
+    partitioned_by: str = None
```
Comment (mrmasterplan) [inline src/atc/configurator/_parse_sql.py:47]: completely unpack the values

File: src/atc/configurator/_parse_sql.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        val: str = token.value.strip()
+        if not _magic_comment_start_re.match(val):
+            continue
+        collected_lines.append(val[len(_magic_comment_start) :])
+    yaml_document = "\n".join(collected_lines)
+    return yaml.load(yaml_document, Loader=yaml.FullLoader) or {}
+
+
+@dataclass
+class StatementBlocks:
```
Comment (LauJohansson) [inline src/atc/configurator/_parse_sql.py:43]: Does it also take delta live tables into account?

File: src/atc/configurator/sql/__init__.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
+from .sqlparse import engine, keywords, tokens
+from .sqlparse.engine import grouping
+from .sqlparse.lexer import Lexer
+
+lex = Lexer()
+lex.clear()
+
+my_regex = [
+    # (r"ZORDER\s+BY\b", tokens.Keyword),
+    (r"IF\s+NOT\s+EXISTS\b", tokens.Keyword),
```
Comment (LauJohansson) [inline src/atc/configurator/sql/__init__.py:10]: Delta live tables?

File: src/atc/configurator/configurator.py [hunk 4]
```diff
@@ -11,6 +10,8 @@
 
 from atc.exceptions import NoSuchValueException
```
Comment (LauJohansson) [inline src/atc/configurator/configurator.py:12]: If the sql parser is dependent on a spark version - maybe we need to assert on it?
I guess some of the SQL syntax only applies for some versions.

File: src/atc/configurator/sql/__init__.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
+from .sqlparse import engine, keywords, tokens
+from .sqlparse.engine import grouping
+from .sqlparse.lexer import Lexer
+
+lex = Lexer()
+lex.clear()
+
+my_regex = [
+    # (r"ZORDER\s+BY\b", tokens.Keyword),
```
Comment (mrmasterplan) [inline src/atc/configurator/sql/__init__.py:9]: copied line

File: src/atc/configurator/_parse_sql.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    partitioned_by: str = None
+    clustered_by: str = None
+    sorted_by: str = None
+    location: str = None
+    comment: str = None
+    tblproperties: str = None
+    as_: str = None
+
+
+def _extract_optional_blocks(stmt) -> StatementBlocks:
```
Comment (LauJohansson) [inline src/atc/configurator/_parse_sql.py:56]: input datatype missing

File: src/atc/configurator/_parse_sql.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+            if not into_token.value.upper() == "INTO":
+                AtcConfiguratorInvalidSqlException("expected INTO after clustering")
+
+            num_buckets = int(next(stmt).value)
+
+            buckets_token = next(stmt)
+            if not buckets_token.value.upper() == "BUCKETS":
+                AtcConfiguratorInvalidSqlException("expected BUCKETS after INTO n")
+
+            cluster_line += f"\nINTO {num_buckets} BUCKETS"
```
Comment (mrmasterplan) [inline src/atc/configurator/_parse_sql.py:114]: forgot to save the wole thing into blocks

File: src/atc/configurator/_parse_sql.py [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+        elif val == ";":
+            break
+        else:
+            raise AtcConfiguratorInvalidSqlException(
+                "Unknown statement form encountered."
+            )
+    return blocks
+
+
+def _walk_create_statement(statement) -> Dict:
```
Comment (LauJohansson) [inline src/atc/configurator/_parse_sql.py:137]: input datatype missing

File: src/atc/configurator/sql/__init__.py (hunk 3 above)
Comment (mrmasterplan) [reply src/atc/configurator/sql/__init__.py:10]: not supported yet

File: src/atc/configurator/_parse_sql.py (hunk 6 above)
Comment (mrmasterplan) [reply src/atc/configurator/_parse_sql.py]: implemented

File: src/atc/configurator/_parse_sql.py (hunk 8 above)
Comment (mrmasterplan) [reply src/atc/configurator/_parse_sql.py]: added

File: src/atc/configurator/_parse_sql.py (hunk 7 above)
Comment (mrmasterplan) [reply src/atc/configurator/_parse_sql.py]: all blocks now implemented and unit-tested

File: src/atc/configurator/_parse_sql.py (hunk 1 above)
Comment (mrmasterplan) [reply src/atc/configurator/_parse_sql.py]: all unpacked now

**Review by LauJohansson — APPROVED** [review]

**Pull request merged**